Implementation of Mempool Syncer

Eligioo opened the pull request:
## What's in this pull request?

The Mempool Syncer is responsible for acquiring the transactions that are currently in the mempool of other nodes directly after reaching consensus.

The mempool syncers get initiated once the Mempool Executors are spawned. One Syncer per Executor is started in order to sync the regular and control transactions separately. First the Syncer starts to discover which transaction hashes other peers with a mempool have and compare those with what we have locally. Then we distribute those unknown hashes among the peers we know have those transactions and retrieve the actual transaction. Lastly for every received transaction we do a full verification which should add it to our local mempool if everything checks out.

## Open design questions
- [x] Define the parameters for the Mempool Syncer: max transactions to sync, transactions per request, etc.
        **Max transactions we acquire per peer: 25.000
        Amount of transaction per request: 500**
- [x] Should the Mempool Syncer end its stream after x amount of time
        **After 10 minutes**

#### This fixes #2625 .

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

Eligioo committed:
Mempool syncer: implementation.

The syncers get initiated once the Mempool Executors are spawned. One Syncer per Executor is started in order to sync the regular and control transactions separately.
First the Syncer starts to discover which transaction hashes other nodes with a mempool have and compare those with what we have locally.
Then we distribute those unknown hashes among the peers we know have those transactions and retrieve the actual transactions.
Lastly for every received transaction we do a full verification which should add it to our local mempool if everything checks out.

hrxi commented on mempool/src/executor.rs:
```diff
@@ -114,16 +114,18 @@ impl<N: Network, T: Topic + Unpin + Sync> Future for MempoolExecutor<N, T> {
-                };
-
-                network.validate_message::<T>(pubsub_id, acceptance);
+                if let Some(pubsub_id) = pubsub_id {
```
We should still somehow punish the sender of an invalid transaction, even if it's not sent via gossipsub.

Eligioo replied:
Added this. If some of the verification variants are encountered, the peer is banned.

Eligioo committed:
Utils: extend FuturesOrdered and FuturesUnordered wrappers with `extend` method

Eligioo committed:
Mempool syncer: organize tests

Eligioo committed:
Mempool syncer: prevent flooding peers with requests

Eligioo committed:
Consensus: mechanism to be able to subscribe for consensus sync events

Eligioo committed:
Mempool task: subscribe to consensus sync and network events to maintain a collection of peers that are in live sync

Eligioo committed:
Mempool syncer: spawn request handlers regardless of whether the mempool syncer is active

Eligioo committed:
Mempool syncer: functionality to add peers to discover its mempool

Eligioo committed:
Consensus: acquire the consensus sync event subscription through the ConsensusProxy rather than the Consensus directly

Eligioo committed:
Test utils: move consensus constructor into the `test_utils` crate

Eligioo committed:
Mempool syncer: subscribe to consensus sync events in order add new peers dynamically

Eligioo committed:
Mempool syncer: minor improvements
- Only try to construct mempool transaction requests if new hashes are discovered
- Check the mempool state first before checking the blockchain if a transaction is already known/included
- Bound the amount of received hashes that will be processed per peer
- Fix tests

Eligioo committed:
Mempool syncer: check peer for necessary service flag when adding it dynamically

sisou commented:
Why is this a breaking change (`testnet-restart` label)?

Eligioo commented:
> Why is this a breaking change (`testnet-restart` label)?

Because it introduces some new network request messages. Technically it is behind a feature flag because the syncer utilizes an yet unused service flag so these new request messages would only go to peers that also have this new component. So even though it's breaking, it shouldn't be too difficult to get in but for bookkeeping I kept the label around.

Eligioo committed:
Mempool syncer: fix merge conflicts

hrxi commented on consensus/src/sync/syncer.rs:
```diff
@@ -16,6 +16,7 @@ use nimiq_network_interface::network::{CloseReason, Network, NetworkEvent, Subsc
 use nimiq_primitives::policy::Policy;
 use nimiq_time::{interval, Interval};
 use nimiq_utils::stream::FuturesUnordered;
+use tokio::sync::broadcast::{channel as broadcast, Sender as BroadcastSender};
```
```suggestion
use tokio::sync::broadcast;
```
And then use `broadcast::channel()` and `broadcast::Sender`. We use it this way in the rest of the code, but I don't know how to enforce that with a clippy lint…

hrxi commented on consensus/src/sync/syncer.rs:
```diff
@@ -159,12 +168,15 @@ impl<N: Network, M: MacroSync<N::PeerId>, L: LiveSync<N>> Syncer<N, M, L> {
             blockchain,
             network,
             network_events,
+            events: tx,
```
```suggestion
            events: broadcast::Sender::new(256),
```

hrxi commented on consensus/src/sync/syncer.rs:
```diff
@@ -185,12 +197,17 @@ impl<N: Network, M: MacroSync<N::PeerId>, L: LiveSync<N>> Syncer<N, M, L> {
     pub fn move_peer_into_live_sync(&mut self, peer_id: N::PeerId) {
         debug!(%peer_id, "Adding peer to live sync");
         self.live_sync.add_peer(peer_id);
+        self.events.send(SyncEvent::AddLiveSync(peer_id)).ok();
```
```suggestion
        let _ = self.events.send(SyncEvent::AddLiveSync(peer_id));
```

Using `.ok()` doesn't communicate that you want to ignore the error. `.ok()` is used to convert the `Ok` case of a `Result<T, E>` into an `Option<T>`.

hrxi commented on consensus/src/sync/syncer.rs:
```diff
@@ -185,12 +197,17 @@ impl<N: Network, M: MacroSync<N::PeerId>, L: LiveSync<N>> Syncer<N, M, L> {
+    pub fn broadcast_sender(&self) -> BroadcastSender<SyncEvent<<N as Network>::PeerId>> {
+        self.events.clone()
+    }
+
```
```suggestion
```

Seems unused. Should probably get a better name if it was used, maybe `events_sender`.

hrxi commented on consensus/src/sync/syncer_proxy.rs:
```diff
@@ -227,6 +228,11 @@ impl<N: Network> SyncerProxy<N> {
+    pub fn broadcast_sender(&self) -> BroadcastSender<SyncEvent<<N as Network>::PeerId>> {
+        gen_syncer_match!(self, broadcast_sender)
+    }
+
```
Unused, too.

hrxi commented on mempool/mempool-task/src/lib.rs:
```diff
@@ -167,12 +195,39 @@ impl<N: Network> MempoolTask<N> {
+        // Process consensus sync updates.
+        while let Poll::Ready(Some(Ok(event))) = self.sync_event_rx.poll_next_unpin(cx) {
+            self.on_consensus_sync_event(event)
+        }
```
This should handle errors, but currently ignores them entirely.

hrxi commented on mempool/mempool-task/src/lib.rs:
```diff
@@ -167,12 +195,39 @@ impl<N: Network> MempoolTask<N> {
+        // Process network updates.
+        while let Poll::Ready(Some(Ok(event))) = self.network_event_rx.poll_next_unpin(cx) {
+            self.on_network_event(event)
+        }
```
This should handle errors as well.

hrxi commented on mempool/src/executor.rs:
```diff
@@ -113,16 +123,37 @@ impl<N: Network, T: Topic + Unpin + Sync> Future for MempoolExecutor<N, T> {
+                        Ok(_) => (),
+                        Err(VerifyErr::InvalidAccount(_))
+                        | Err(VerifyErr::InvalidBlockNumber)
+                        | Err(VerifyErr::InvalidTransaction(_)) => {
```
What's the reasoning for including specifically these errors for banning the client and not the others?

Can a transaction become invalid with one of these errors while we're requesting them because we apply a new block?

It'd be nice to have a comment above to explain which error types were chosen.

Eligioo replied:
If you look at all the variants of [`VerifyErr`](https://github.com/nimiq/core-rs-albatross/blob/8280e813459a1ee51c1a53d3363396aff9a18765/mempool/src/verify.rs#L16-L31), not all of them should result into banning the peer. So I picked the ones that are considered malicious and will cause into a ban.

I also unified the `pubsub_or_peer_id` and made the match on the `VerifyErr` exhaustive in the case.

Added comments to each variant

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+/// store if we have a requested the corresponding transaction already to one of those peers.
+struct HashRequestStatus<N: Network> {
+    status: RequestStatus,
+    peer_ids: Vec<N::PeerId>,
```
```suggestion
    peer_ids: HashSet<N::PeerId>,
```
Looks like this wants fast lookup, so a `HashSet` would be more appropriate.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+    /// Timeout to gracefully shutdown the mempool syncer entirely
+    shutdown_timer: Pin<Box<Sleep>>,
+
+    /// Consensus sync event receiver
```
```suggestion
```
The comment doesn't tell me what it does, can be omitted.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+    /// Consensus sync event receiver
+    consensus_sync_event_rx: BroadcastStream<SyncEvent<<N>::PeerId>>,
+
+    /// Blockchain reference
```
```suggestion
```

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+    >,
+
+    /// Collection of transaction hashes not present in the local mempool
+    unknown_hashes: HashMap<Blake2bHash, HashRequestStatus<N>>,
```
I don't see anything removing transactions from this hashmap.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+        let hashes_requests = peers
+            .iter()
+            .map(|peer_id| {
+                let peer_id = *peer_id;
```
```suggestion
            .map(|&peer_id| {
```

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+        debug!(peer_id = %peer_id, num = %hashes.len(), "Received unknown mempool hashes");
+        let mut new_hashes_discovered = false;
+
+        hashes.into_iter().take(MAX_TOTAL_HASHES).for_each(|hash| {
```
What is the `.take(MAX_TOTAL_HASHES)` supposed to do here?

Eligioo replied:
To bound the amount of hashes we receive and process from the peers we are requesting them from. On the server side it's also already capped by the `MAX_TOTAL_HASHES`. Those hashes can even be fake so I wanted it to limit it some way.

hrxi replied:
So should it be `MAX_HASHES_PER_REQUEST`?

Eligioo replied:
No, since we want to receive at most `MAX_TOTAL_HASHES` (25.000) hashes from a peer when requesting it mempool hashes.

`MAX_HASHES_PER_REQUEST` is the constant to limit the amount of the actual transactions we ask for per request. So for example we want to request 2000 transactions in total,  `MAX_HASHES_PER_REQUEST` makes sure those transactions are batched into 4 separate requests each containing `MAX_HASHES_PER_REQUEST` transactions.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+        debug!(peer_id = %peer_id, num = %hashes.len(), "Received unknown mempool hashes");
+        let mut new_hashes_discovered = false;
+
+        hashes.into_iter().take(MAX_TOTAL_HASHES).for_each(|hash| {
```
`for` loop would probably be more readable here, rather than `for_each(…)`

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+                    entry.insert(HashRequestStatus::new(vec![peer_id]));
+                    new_hashes_discovered = true;
+                }
+            };
```
```suggestion
            match self.unknown_hashes.entry(hash).or_insert_with(|| {
                new_hashes_discovered = true;
                HashRequestStatus::new(vec![])
            }).add_peer(peer_id);
```

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+        if self.peers.contains(&peer_id)
+            || !self
+                .network
+                .peer_provides_services(peer_id, Services::MEMPOOL)
```
Since the mempool syncer is doing the filtering here, it should probably also do that in the constructor, rather than the caller doing that.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+    pub fn mark_as_received(&mut self) {
+        self.status = RequestStatus::Received
+    }
+}
```
Since the members are also accessed directly, maybe we can skip all these member functions and operate directly on the members.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+            }
+        }
+
+        let requests = prepared_requests.into_iter().map(|request| {
```
```suggestion
        let requests = prepared_requests.into_iter().map(|(request, peer_id)| {
```

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+        {
+            for i in 0..self.peers.len() {
+                let peer = self.peers[i];
+                if let Some(request) = self.batch_hashes_by_peer_id(peer) {
```
This looks like O(n²) over the `self.unknown_hashes`. Maybe it would be possible to keep unrequested hashes in a different collection than requested hashes so that we don't have to filter all the time?

I guess one could also invert this loop — for each unknown hash, pick one of the possible peers and add it to that peer's list.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+            return Poll::Ready(None);
+        }
+
+        // Then we check if peers got added to live sync in the mean time
```
```suggestion
        // Then we check if peers got added to live sync in the meantime
```
"meantime" has fewer meanings than "mean time".

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+                }
+                Err(err) => {
+                    error!(%err, %peer_id, "Failed to fetch mempool transactions");
+                }
```
We should probably re-request these transactions from another peer.

hrxi commented on mempool/src/sync/mod.rs:
```diff
@@ -0,0 +1,421 @@
+                    }
+
+                    info!(num = %transactions.len(), "Synced mempool transactions");
+                    self.transactions.extend(transactions);
```
What about the response missing transactions that we requested?

Eligioo replied:
I my opinion that's OK when you don't receive (all) transactions you are asking for. By the time the peer received your request, the particular transaction(s) could already be adopted in a block.

hrxi commented on mempool/src/executor.rs:
```diff
@@ -113,16 +123,37 @@ impl<N: Network, T: Topic + Unpin + Sync> Future for MempoolExecutor<N, T> {
+                        };
+
+                        network.validate_message::<T>(pubsub_id, acceptance);
+                    }
```
Can we somehow unify these branches so that we don't have duplication on what errors we consider bad?

hrxi commented:
> > Why is this a breaking change (`testnet-restart` label)?
> 
> Because it introduces some new network request messages. Technically it is behind a feature flag because the syncer utilizes an yet unused service flag so these new request messages would only go to peers that also have this new component. So even though it's breaking, it shouldn't be too difficult to get in but for bookkeeping I kept the label around.

I think I agree with @sisou that it's not breaking, precisely because it uses a previously unused feature flag.

Eligioo committed:
Apply PR feedback

Eligioo committed:
Fix merge conflicts

Eligioo commented:
@hrxi either your comments are resolved or has received a reply
